### ISSUE TITLE
500 on download of H5Ps

### ISSUE THREAD

**Issue by tmkarlsen** [cerpus/Edlib#2326]
**Reuse download gives 500 error**

500 error when trying to download https://www.h5p.ndla.no/s/resources/a2e447e2-2e7f-4b31-8e13-3ef2429173a1.

Could be due to large video file and shortage of memory or issues with subtitles

**Comment by emmachughes** [comment]
After investigating, we've seen multiple issues that can lead to exporting videos failing:

1. **Out of memory**

   This may or may not be fixed by #2309.

2. **Text track URLs from Brightcove are mangled when inserting videos**

   When exporting videos, stuff like text tracks will be downloaded and included in the export.

   The Brightcove API gives us text track URLs like these:
   
   `http://house-fastly-signed-eu-west-1-prod.brightcovecdn.com/media/v1/text/vtt/clear/4806596774001/9629d900-3e88-4a5b-987d-4cd07ed29fd7/ee1d525d-22e0-4369-868b-856191cb6319/text.vtt?fastly_token=NjQ3Zjc4ZmZfODQ5NDY2MDk5MmI0OWYxMTFmYWUzNzhlNWRiZDYyOTdjZmFhMTUxYjU0OGM0ZDI2MzIwM2ExZDA5ODVhOGE1ZV9odHRwOi8vaG91c2UtZmFzdGx5LXNpZ25lZC1ldS13ZXN0LTEtcHJvZC5icmlnaHRjb3ZlY2RuLmNvbS9tZWRpYS92MS90ZXh0L3Z0dC9jbGVhci80ODA2NTk2Nzc0MDAxLzk2MjlkOTAwLTNlODgtNGE1Yi05ODdkLTRjZDA3ZWQyOWZkNy9lZTFkNTI1ZC0yMmUwLTQzNjktODY4Yi04NTYxOTFjYjYzMTkvdGV4dC52dHQ%3D`

   This is a signed URL, so requires a valid signature or we cannot download the text track. Decoding the long base64 string, we see there's an expiry timestamp (hex-encoded), HMAC, and the URL without the query string:
   
   `647f78ff_8494660992b49f111fae378e5dbd6297cfaa151b548c4d263203a1d0985a8a5e_http://house-fastly-signed-eu-west-1-prod.brightcovecdn.com/media/v1/text/vtt/clear/4806596774001/9629d900-3e88-4a5b-987d-4cd07ed29fd7/ee1d525d-22e0-4369-868b-856191cb6319/text.vtt`

   Our Brightcove integration [will naively replace the string `http` with `https` in text track URLs](https://github.com/cerpus/Edlib/blob/999acd0b33d691f33d0dda86330510fd0b2e32fe/sourcecode/apis/contentauthor/public/js/h5p/ndla-contentbrowser.js#L535). Since the URL no longer matches the encoded one used to generate the HMAC, Brightcove rejects the request, and exporting fails.

3. **Text track URLs expire**

   Since the Brightcove URLs have an expiry date, attempting to export an H5P with a text track URL that expired will fail.

4. **Text track URLs with `https` schemes are mangled**

   The aforementioned substitution of `http` with `https` will cause URL with the `https://` scheme will be mangled to become `httpss://`. This has caused URLs like `https://statisk.ndla.no/videotekster/6160288014001.nb-NO.vtt` to be replaced with `httpss://statisk.ndla.no/videotekster/6160288014001.nb-NO.vtt`.

In conclusion, this is a big ducking mess.

To fix this, we must:

1. Ensure there are no more out-of-memory errors.
 
2. Remove the mangling of URLs, however this may cause mixed http/https errors when NDLA loads text tracks. Presumably, the substitution was done for a reason.

3. Unmangle existing URLs. These are in CA's `h5p_contents.parameters` column, which is JSON-encoded, so will not be straight forward to fix.

4. Find an alternative to the text track URLs that doesn't expire. For this, we might need to proxy Brightcove's text tracks behind our own endpoint.

**Comment by tmkarlsen** [comment]
Brightcove API documentation: https://apis.support.brightcove.com/index.html
